File: devops-agent-iac/abordagem-2-cdk-python/app_lib/stacks/data_stack.py
"""Camada de armazenamento consumida por Glue, Athena, Lambda e Step Functions."""
import aws_cdk as cdk
from aws_cdk import aws_kms as kms
from aws_cdk import aws_s3 as s3
from constructs import Construct

from app_lib.config import EnvironmentConfig, prefix
from app_lib.constructs.secure_bucket import SecureBucket


class DataStack(cdk.Stack):
    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        *,
        env_name: str,
        project_name: str,
        config: EnvironmentConfig,
        data_key: kms.IKey,
        **kwargs,
    ) -> None:
        super().__init__(scope, construct_id, **kwargs)
        name = prefix(project_name, env_name)
        is_prod = env_name == "prod"
        _ = config

        self.raw_bucket = SecureBucket(
            self,
            "Raw",
            bucket_name=f"{name}-py-raw-{self.account}",
            encryption_key=data_key,
            retain=is_prod,
            lifecycle_rules=[
                s3.LifecycleRule(
                    id="tiering",
                    enabled=True,
                    transitions=[
                        s3.Transition(
                            storage_class=s3.StorageClass.INTELLIGENT_TIERING,
                            transition_after=cdk.Duration.days(30),
                        ),
                        s3.Transition(
                            storage_class=s3.StorageClass.GLACIER_INSTANT_RETRIEVAL,
                            transition_after=cdk.Duration.days(180),
                        ),
                    ],
                    noncurrent_version_expiration=cdk.Duration.days(90),
                )
            ],
        ).bucket

        self.curated_bucket = SecureBucket(
            self,
            "Curated",
            bucket_name=f"{name}-py-curated-{self.account}",
            encryption_key=data_key,
            retain=is_prod,
        ).bucket

        self.athena_results_bucket = SecureBucket(
            self,
            "AthenaResults",
            bucket_name=f"{name}-py-athena-results-{self.account}",
            encryption_key=data_key,
            retain=False,
            lifecycle_rules=[
                s3.LifecycleRule(id="expire-results", enabled=True, expiration=cdk.Duration.days(30))
            ],
        ).bucket

File: devops-agent-iac/anexos/cdk-typescript/lib/data-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as kms from 'aws-cdk-lib/aws-kms';
import { Construct } from 'constructs';
import { BaseProps, prefix } from './config';
import { SecureBucket } from './constructs/secure-bucket';

export interface DataStackProps extends cdk.StackProps, BaseProps {
  readonly dataKey: kms.IKey;
}

/** Camada de armazenamento que Glue, Athena, Lambda e Step Functions consumirão. */
export class DataStack extends cdk.Stack {
  public readonly rawBucket: s3.IBucket;
  public readonly curatedBucket: s3.IBucket;
  public readonly athenaResultsBucket: s3.IBucket;

  constructor(scope: Construct, id: string, props: DataStackProps) {
    super(scope, id, props);
    const p = prefix(props);
    const isProd = props.envName === 'prod';

    this.rawBucket = new SecureBucket(this, 'Raw', {
      bucketName: `${p}-raw-${this.account}`,
      encryptionKey: props.dataKey,
      retain: isProd,
      lifecycleRules: [
        {
          id: 'tiering',
          enabled: true,
          transitions: [
            { storageClass: s3.StorageClass.INTELLIGENT_TIERING, transitionAfter: cdk.Duration.days(30) },
            { storageClass: s3.StorageClass.GLACIER_INSTANT_RETRIEVAL, transitionAfter: cdk.Duration.days(180) },
          ],
          noncurrentVersionExpiration: cdk.Duration.days(90),
        },
      ],
    }).bucket;

    this.curatedBucket = new SecureBucket(this, 'Curated', {
      bucketName: `${p}-curated-${this.account}`,
      encryptionKey: props.dataKey,
      retain: isProd,
    }).bucket;

    this.athenaResultsBucket = new SecureBucket(this, 'AthenaResults', {
      bucketName: `${p}-athena-results-${this.account}`,
      encryptionKey: props.dataKey,
      retain: false,
      lifecycleRules: [{ id: 'expire-results', enabled: true, expiration: cdk.Duration.days(30) }],
    }).bucket;
  }
}
